addons/product/models/product_attribute.py

```diff
@@ -526,6 +533,31 @@ def _is_from_single_value_line(self, only_active=True):
             all_values = all_values._only_active()
         return len(all_values) == 1
 
+    def _unarchive_variants(self):
+        """
+        Unarchive all the variants for product_template
+        """
+        variants = self.product_tmpl_id.product_variant_ids.with_context(active_test=False).product_variant_ids
+        for variant in variants:
+            variant.active = True
+
+    def _archive_variants_with_exclusion(self):
```

Comment:
Isn't it possible to do compute it before `write`? It's strange to let those be un-archived in the first place.

Reply:
Anything variant related should be handled in `_create_variant_ids`, not split concern in multiple places.

Reply:
> Anything variant related should be handled in `_create_variant_ids`, not split concern in multiple places.

I'll try to figure out how I can keep this logic in `_create_variant_ids` and change the flow so it only unarchives/archives when necessary.

Thank you for the feedback.

Reply:
Note that `_create_variant_ids` is already doing operations only when necessary, so it shouldn't change much.
What you'd want here is to simply call it in `write` when exclusion changes.

And slightly adapt it to not "incorrectly" put stuff in `variants_to_activate` or `variants_to_create` when there are exclusions, but into `variants_to_unlink` (if existing).

"Worst case" scenario we could probably just iterate the exclusions to "post-filter" these 3 lists just before the operations are done. (just make sure whatever you do is still fast if there are many combination/variants existing)